### DISCUSSION TITLE
Truenas FreeBSD package?

### DISCUSSION THREAD

**Discussion by Godfear17** [dswd/vpncloud#175] [Q&A]
Is there a package for FreeBSD? want to install this on a Truenas core server which is FreeBSD to join my other vpn mesh servers. 

Or is there a way to install this on FreeBSD if there isn't a package?

**Comment by dswd** [comment]
Right now there is no FreeBSD version available. Currently the code does not compile for FreeBSD. I am working on some changes that might make it compile (as a byproduct) but that might take some time.

**Comment by Godfear17** [comment]
oof, that's bad news for Truenas servers.

**Reply by windwardly** [reply]
Yes, unfortunate, and too, an opportunity to push a little. I keep a "build" jail on my truenas box, and have some compilers, git, screen, and ssh server for things like this. I ssh in, git clone, cargo build, see errors (no `WaitImpl` in `poll`; libc::SIOCGIFNETMASK: not found in `libc`, ...). At any step I may shrug -- more effort to keep going than I'm willing to muster in this moment. But if each time I push a little, then little by little, my jail becomes more useful and I may eventually get something working when it didn't. All this started with "I should set up a jail; but then I'd have to learn how to do that" and one day, I learned how to do that. "I should set it up so I could ssh in; but that's too much", and then one day, I figured that out. "I should install Rust".. you get the idea.

That said, I scroll through the errors, and see that they are libc/syscall related things, which would take more time to understand than I'd like in this moment. But it did lead me to install asciidoc in my build jail. Little by little.

**Reply by dswd** [reply]
Yes, so the main problem here is that BSD is not Linux. This hits you when it comes to the details about TUN/TAP interfaces, epoll and configuring network interfaces.
* The TUN/TAP part should be solved already
* The epoll is quite hard. I use it to wait on the network device and the udp port at the same time, because VpnCloud is single-threaded. However I am working on multi-threading for some months now (this is complex) and already have something that compiles and passes most tests. I think it will still take some months to get it ready but once I have it ready, there should be no need anymore for epoll/kqueue.
* Setting up devices is nice to have but not essential for the core functionality. Either we get that from the BSD libc or we call a system command (ifconfig ?).

One core problem is that I don't have any BSD system to test and develop on and frankly I don't have time to work on multiple platforms at the same time. The multithreading is already taking way too long on one platform. It would be great if someone can push the development on BSD.
